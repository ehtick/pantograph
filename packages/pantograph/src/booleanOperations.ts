import {
  cutFiguresLists,
  fuseFiguresLists,
  intersectFiguresLists,
} from "./algorithms/boolean/figureBooleans";
import { Diagram } from "./models/Diagram";
import { Figure } from "./models/Figure";
import { Loop } from "./models/Loop";
import { listOfFigures } from "./utils/listOfFigures";

export function fuse(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    fuseFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}

export function fuseAll(shapes: (Diagram | Figure | Loop)[]): Diagram {
  return shapes.reduce(
    (acc: Diagram, shape: Diagram | Figure | Loop) => fuse(acc, shape),
    new Diagram()
  );
}

export function cut(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    cutFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}

export function intersect(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    intersectFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}
