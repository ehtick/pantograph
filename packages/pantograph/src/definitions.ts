export type Vector = [number, number];

export function isVector(vector: unknown): vector is Vector {
  return Array.isArray(vector) && vector.length === 2;
}
