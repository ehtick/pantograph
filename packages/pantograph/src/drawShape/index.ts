export { drawCircle } from "./drawCircle.js";
export { drawRect } from "./drawRect.js";
export { drawEllipse } from "./drawEllipse.js";
export { drawSVGPath } from "./drawSVGPath.js";
