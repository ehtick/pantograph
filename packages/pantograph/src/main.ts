import type { Vector } from "./definitions";
export { Vector };

export const DEG2RAD = Math.PI / 180;
export const RAD2DEG = 180 / Math.PI;

export function polarToCartesian(r: number, theta: number): Vector {
  const x = Math.cos(theta * DEG2RAD) * r;
  const y = Math.sin(theta * DEG2RAD) * r;
  return [x, y];
}

export function cartesianToPolar([x, y]: Vector): [number, number] {
  const r = Math.sqrt(x * x + y * y);
  const theta = Math.atan2(y, x) * RAD2DEG;
  return [r, theta];
}

export type {
  Diagram,
  Figure,
  Loop,
  Strand,
  Stroke,
  TransformationMatrix,
  BoundingBox,
  Segment,
  Line,
  Arc,
} from "./models/exports";

export { draw } from "./draw";
export {
  // Surface booleans
  fuseAll,
  fuse,
  cut,
  intersect,
  // Strand booleans
  eraseStrand,
  confineStrand,
  // Offset
  offset,
} from "./operations";

export { exportSVG, svgBody } from "./export/svg/exportSVG";

export { exportJSON } from "./export/json/exportJSON";
export { importJSON } from "./import/json/importJSON";
