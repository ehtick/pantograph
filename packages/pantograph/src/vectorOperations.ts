import type { Vector } from "./definitions";

export const DEG2RAD = Math.PI / 180;

export const reprVector = (vector: Vector): string => {
  return `[${vector[0]}, ${vector[1]}]`;
};

export const sameVector = (
  [x0, y0]: Vector,
  [x1, y1]: Vector,
  precision = 1e-9
): boolean => {
  return Math.abs(x0 - x1) <= precision && Math.abs(y0 - y1) <= precision;
};

export const add = ([x0, y0]: Vector, [x1, y1]: Vector): Vector => {
  return [x0 + x1, y0 + y1];
};

export const subtract = ([x0, y0]: Vector, [x1, y1]: Vector): Vector => {
  return [x0 - x1, y0 - y1];
};

export const squareLength = ([x, y]: Vector): number => {
  return x * x + y * y;
};

export const length = ([x, y]: Vector): number => {
  return Math.sqrt(squareLength([x, y]));
};

export const scalarMultiply = ([x0, y0]: Vector, scalar: number): Vector => {
  return [x0 * scalar, y0 * scalar];
};

export const squareDistance = (
  [x0, y0]: Vector,
  [x1, y1]: Vector = [0, 0]
): number => {
  return (x0 - x1) ** 2 + (y0 - y1) ** 2;
};

export const distance = (p0: Vector, p1: Vector = [0, 0]): number => {
  return Math.sqrt(squareDistance(p0, p1));
};

export function crossProduct([x0, y0]: Vector, [x1, y1]: Vector): number {
  return x0 * y1 - y0 * x1;
}

export function dotProduct([x0, y0]: Vector, [x1, y1]: Vector): number {
  return x0 * x1 + y0 * y1;
}

export const angle = ([x0, y0]: Vector, [x1, y1]: Vector = [0, 0]): number => {
  return Math.atan2(y1 * x0 - y0 * x1, x0 * x1 + y0 * y1);
};

export function normalize([x0, y0]: Vector): Vector {
  const l = distance([x0, y0]);
  return [x0 / l, y0 / l];
}

export function polarToCartesian(r: number, theta: number): Vector {
  const x = Math.cos(theta) * r;
  const y = Math.sin(theta) * r;
  return [x, y];
}

export function cartesianToPolar([x, y]: Vector): [number, number] {
  const r = distance([x, y]);
  const theta = Math.atan2(y, x);

  return [r, theta];
}

export function parallel(v1: Vector, v2: Vector, precision = 1e-9): boolean {
  const V1xV2 = crossProduct(v1, v2);

  const xLength = squareLength(v1);
  const yLength = squareLength(v2);

  return V1xV2 * V1xV2 < xLength * yLength * precision * precision;
}
