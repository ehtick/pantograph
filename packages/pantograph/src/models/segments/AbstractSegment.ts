import type { Vector } from "../../definitions";
import { BoundingBox } from "../BoundingBox";
import type { TransformationMatrix } from "../TransformationMatrix";
import { reprVector } from "../../vectorOperations";
import { Transformable } from "../utils/Transformable";

export abstract class AbstractSegment<
  T extends AbstractSegment<T>,
> extends Transformable<T> {
  constructor(
    public firstPoint: Vector,
    public lastPoint: Vector,
  ) {
    super();
    this.firstPoint = firstPoint;
    this.lastPoint = lastPoint;
  }

  readonly precision: number = 1e-9;

  abstract segmentType: string;

  get repr() {
    return `${this.segmentType} ${reprVector(this.firstPoint)} - ${reprVector(
      this.lastPoint,
    )}`;
  }
  get info() {
    return this.repr;
  }

  abstract get midPoint(): Vector;

  abstract get boundingBox(): BoundingBox;

  abstract clone(): T;

  abstract reverse(): T;

  abstract isSame(other: AbstractSegment<any>): boolean;

  abstract distanceFrom(element: Vector): number;

  abstract isOnSegment(point: Vector): boolean;

  abstract tangentAt(point: Vector): Vector;

  abstract get tangentAtFirstPoint(): Vector;
  abstract get tangentAtLastPoint(): Vector;

  abstract splitAt(points: Vector[] | number[]): T[];

  abstract transform(matrix: TransformationMatrix): T;
}
