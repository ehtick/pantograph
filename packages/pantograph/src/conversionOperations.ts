import { Segment } from "./models/segments/Segment";
import { Arc } from "./models/segments/Arc";
import { Line } from "./models/segments/Line";
import { QuadraticBezier } from "./models/segments/QuadraticBezier";
import { CubicBezier } from "./models/segments/CubicBezier";
import { EllipseArc } from "./models/segments/EllipseArc";
import { approximateBezierCurveAsArcs } from "./algorithms/conversions/bezierToArcs";
import { approximateEllipticalArcAsCubicBeziers } from "./algorithms/conversions/ellipseToBezier";
import { genericConversion } from "./algorithms/conversions/helpers";
import { Diagram, Figure, Loop, Strand } from "./main";

function mapAsArcsAndLines(s: Segment, tolerance = 1e-3): Segment[] {
  if (s instanceof Arc || s instanceof Line) {
    return [s];
  }
  if (s instanceof QuadraticBezier || s instanceof CubicBezier) {
    return approximateBezierCurveAsArcs(s, tolerance);
  }
  if (s instanceof EllipseArc) {
    return approximateEllipticalArcAsCubicBeziers(s).flatMap((s) =>
      approximateBezierCurveAsArcs(s, tolerance),
    );
  }

  throw new Error(`Unknown segment type: ${s}`);
}

function convertToArcsAndLines(shape: Diagram): Diagram;
function convertToArcsAndLines(shape: Figure): Figure;
function convertToArcsAndLines(shape: Loop): Loop;
function convertToArcsAndLines(shape: Strand): Strand;
function convertToArcsAndLines(shape: Segment): Segment[];
function convertToArcsAndLines(
  shape: Diagram | Figure | Loop | Strand | Segment,
  tolerance = 1e-3,
) {
  const mapper = (segment: Segment) => mapAsArcsAndLines(segment, tolerance);
  return genericConversion(shape, mapper);
}

export { convertToArcsAndLines };
