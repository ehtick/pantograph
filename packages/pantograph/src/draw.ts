import { Vector } from "./definitions";
import { Diagram } from "./models/Diagram";
import { Figure } from "./models/Figure";
import { Loop } from "./models/Loop";
import { Line } from "./models/segments/Line";
import { Segment } from "./models/segments/Segment";
import { TransformationMatrix } from "./models/TransformationMatrix";
import {
  polarToCartesian,
  DEG2RAD,
  subtract,
  sameVector,
} from "./vectorOperations";

function closeSegments(segments: Segment[]) {
  if (!segments.length) throw new Error("No segments to close");
  const firstSegment = segments[0];
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const lastSegment = segments.at(-1)!;

  if (sameVector(firstSegment.firstPoint, lastSegment.lastPoint))
    return segments;
  return [
    ...segments,
    new Line(lastSegment.lastPoint, firstSegment.firstPoint),
  ];
}

function loopySegmentsToDiagram(segments: Segment[]) {
  // Here we will need to do our best to fix cases where the drawing is
  // broken in some way (i.e. self-intersecting loops)

  return new Diagram([new Figure(new Loop(segments))]);
}

export class DrawingPen {
  protected pointer: Vector;
  protected firstPoint: Vector;
  protected pendingSegments: Segment[];

  constructor(origin: Vector = [0, 0]) {
    this.pointer = origin;
    this.firstPoint = origin;

    this.pendingSegments = [];
  }

  movePointerTo(point: Vector): this {
    if (this.pendingSegments.length)
      throw new Error(
        "You can only move the pointer if there is no segment defined"
      );

    this.pointer = point;
    this.firstPoint = point;
    return this;
  }

  protected saveSegment(segment: Segment) {
    this.pendingSegments.push(segment);
    return this;
  }

  lineTo(point: Vector): this {
    const segment = new Line(this.pointer, point);
    this.pointer = point;
    return this.saveSegment(segment);
  }

  line(xDist: number, yDist: number): this {
    return this.lineTo([this.pointer[0] + xDist, this.pointer[1] + yDist]);
  }

  vLine(distance: number): this {
    return this.line(0, distance);
  }

  hLine(distance: number): this {
    return this.line(distance, 0);
  }

  vLineTo(yPos: number): this {
    return this.lineTo([this.pointer[0], yPos]);
  }

  hLineTo(xPos: number): this {
    return this.lineTo([xPos, this.pointer[1]]);
  }

  polarLineTo([r, theta]: Vector): this {
    const angleInRads = theta * DEG2RAD;
    const point = polarToCartesian(r, angleInRads);
    return this.lineTo(point);
  }

  polarLine(distance: number, angle: number): this {
    const angleInRads = angle * DEG2RAD;
    const [x, y] = polarToCartesian(distance, angleInRads);
    return this.line(x, y);
  }

  tangentLine(distance: number): this {
    const previousCurve = this.pendingSegments.at(-1);

    if (!previousCurve)
      throw new Error("You need a previous segment to sketch a tangent line");

    const [xDir, yDir] = previousCurve.tangentAtLastPoint;
    return this.line(xDir * distance, yDir * distance);
  }

  close(): Diagram {
    const segments = closeSegments(this.pendingSegments);
    return loopySegmentsToDiagram(segments);
  }

  closeWithMirror(): Diagram {
    if (!this.pendingSegments.length) throw new Error("No segments to close");

    const firstSegment = this.pendingSegments[0];
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const lastSegment = this.pendingSegments.at(-1)!;

    const mirrorVector = subtract(
      lastSegment.lastPoint,
      firstSegment.firstPoint
    );
    const mirrorTranform = new TransformationMatrix().mirrorLine(
      mirrorVector,
      firstSegment.firstPoint
    );

    const mirroredSegments = this.pendingSegments.map((segment) =>
      segment.transform(mirrorTranform).reverse()
    );
    mirroredSegments.reverse();

    return loopySegmentsToDiagram([
      ...this.pendingSegments,
      ...mirroredSegments,
    ]);
  }
}

export function draw(origin: Vector = [0, 0]): DrawingPen {
  return new DrawingPen(origin);
}
