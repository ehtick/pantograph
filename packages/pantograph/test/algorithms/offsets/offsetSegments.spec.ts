import { describe, it, expect } from "vitest";

import {
  DegenerateSegment,
  offsetArc,
  offsetLine,
} from "../../../src/algorithms/offsets/offsetSegment";
import { Arc, tangentArc } from "../../../src/models/segments/Arc";
import { Line } from "../../../src/models/segments/Line";
import { distance } from "../../../src/vectorOperations";
import { debugImg } from "../../debug";

describe("offsetSegments", () => {
  describe("offsetLine", () => {
    it("offsets a vertical line and its reverse", () => {
      const line = new Line([0, -1], [0, 1]);
      const offset = offsetLine(line, 0.5);

      expect(offset.firstPoint).toEqual([-0.5, -1]);
      expect(offset.lastPoint).toEqual([-0.5, 1]);

      const offset2 = offsetLine(line.reverse(), 0.5);

      expect(offset2.firstPoint).toEqual([0.5, 1]);
      expect(offset2.lastPoint).toEqual([0.5, -1]);
    });

    it("offsets an horizontal line and its reverse", () => {
      const line = new Line([-1, 0], [1, 0]);
      const offset = offsetLine(line, 0.5);

      expect(offset.firstPoint).toEqual([-1, 0.5]);
      expect(offset.lastPoint).toEqual([1, 0.5]);

      const offset2 = offsetLine(line.reverse(), 0.5);

      expect(offset2.firstPoint).toEqual([1, -0.5]);
      expect(offset2.lastPoint).toEqual([-1, -0.5]);
    });

    it("offsets an angled line and its reverse", () => {
      const line = new Line([-1, 2], [1, 3]);
      const offset = offsetLine(line, 1);

      expect(distance(offset.firstPoint, line.firstPoint)).toBeCloseTo(1);
      expect(distance(offset.lastPoint, line.lastPoint)).toBeCloseTo(1);

      const line2 = line.reverse();
      const offset2 = offsetLine(line2, 1.5);

      expect(distance(offset2.firstPoint, line2.firstPoint)).toBeCloseTo(1.5);
      expect(distance(offset2.lastPoint, line2.lastPoint)).toBeCloseTo(1.5);

      expect(distance(offset2.firstPoint, offset.lastPoint)).toBeCloseTo(2.5);
      expect(distance(offset2.lastPoint, offset.firstPoint)).toBeCloseTo(2.5);
    });
  });

  describe("offsetArc", () => {
    it("offsets a counter clockwise arc", () => {
      const arc = new Arc([0, -1], [0, 1], [0, 0]).rotate(5);
      const offset = offsetArc(arc, 0.5) as Arc;

      expect(offset.center).toEqual(arc.center);
      expect(offset.radius).toEqual(arc.radius - 0.5);

      expect(offset.firstAngle).toEqual(arc.firstAngle);
      expect(offset.lastAngle).toEqual(arc.lastAngle);

      expect(offsetArc(arc, -0.5).radius).toBeCloseTo(arc.radius + 0.5);
    });

    it("offsets a clockwise arc", () => {
      const arc = new Arc([0, -1], [0, 1], [0, 0]).reverse().rotate(5);
      const offset = offsetArc(arc, 0.5) as Arc;

      expect(offset.center).toEqual(arc.center);
      expect(offset.radius).toEqual(arc.radius + 0.5);

      expect(offset.firstAngle).toEqual(arc.firstAngle);
      expect(offset.lastAngle).toEqual(arc.lastAngle);

      expect(offsetArc(arc, -0.5).radius).toBeCloseTo(arc.radius - 0.5);
    });

    it("returns the degenerate case", () => {
      const arc = new Arc([0, 1], [1, 0], [0, 0]);
      const offset = offsetArc(arc, 1.5) as Line;

      debugImg([arc, new Line(offset.firstPoint, offset.lastPoint)]);

      expect(offset).toBeInstanceOf(DegenerateSegment);

      expect(offset.firstPoint).toBeVector([0, -0.5]);
      expect(offset.lastPoint).toBeVector([-0.5, 0]);
    });
  });
});
