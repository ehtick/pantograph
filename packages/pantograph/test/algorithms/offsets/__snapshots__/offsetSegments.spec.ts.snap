// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`offsetSegments > offsetBezier > offsets a bezier 1`] = `
[
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      0.580952380952381,
      0.4,
    ],
    "firstPoint": [
      0,
      0.4,
    ],
    "lastPoint": [
      0.8735088935932649,
      0.49058443033131666,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      0.9317546119113013,
      0.5099996697706621,
    ],
    "firstPoint": [
      0.8735088935932649,
      0.49058443033131666,
    ],
    "lastPoint": [
      0.9671572875253809,
      0.5328427124746189,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
  DegenerateSegment {
    "firstPoint": [
      0.9671572875253813,
      0.5328427124746193,
    ],
    "lastPoint": [
      0.9333333333333335,
      0.4444444444444444,
    ],
  },
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      0.9333333333333335,
      0.5666666666666667,
    ],
    "firstPoint": [
      0.9333333333333335,
      0.4444444444444444,
    ],
    "lastPoint": [
      0.7171572875253811,
      0.7171572875253809,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
]
`;
